jlm/rvsdg/NodeNormalization.hpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_NODENORMALIZATION_HPP
+#define JLM_RVSDG_NODENORMALIZATION_HPP
+
+#include <jlm/rvsdg/node.hpp>
+#include <jlm/rvsdg/region.hpp>
+#include <jlm/util/common.hpp>
+
+#include <functional>
+#include <optional>
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+class output;
+
+template<class TOperation>
+using NodeNormalization = std::function<
+    std::optional<std::vector<output *>>(const TOperation &, const std::vector<output *> &)>;
+
+template<class TOperation>
+using NodeNormalizationSequenceType = std::function<NodeNormalization<TOperation>(
+    const std::vector<NodeNormalization<TOperation>> &)>;
+
+template<class TOperation>
+NodeNormalizationSequenceType<TOperation> NodeNormalizationSequence =
```

Comment:
Does `NodeNormalizationSequence` need to be a lambda, and not just a regular function? Will it ever be sent anywhere? It makes sense to use `std::function` for the normalizations themselves, but will this function ever be used in a general setting?

Reply:
The idea for the normalization sequence is to be a normalization itself again. That is why the type takes a sequence of normalizations, which are applied in order until the first one that triggers, and returns a normalization again. This means I can push the normalization sequence in any place where a normal normalization is expected, e.g., NodeReduction.

Reply:
I agree that a sequence of normalizations can itself be seen as a normalization. The signature of the function makes this clear as well. The function I am talking about is the "outer" function, which takes a list of normalizations and produces a single normalization. This outer function is not a normalization itself, so it does not have to go via std::function or lambdas, unless there will be other functions with the signature `(vector<Normalization>) -> Normalization` that we went to use in generic settings.

Reply:
True. I wish there was a nice way of declaring that a function has a signature that conforms to an interface in C++. It is not a big issue, however, as you would get compilation errors as soon as you try converting your function to an `std::function` through the typedef.

Reply:
Converted to plain functions